// @aws-cdk/aws-events-targets/lib/event-bus.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventBus = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("@aws-cdk/aws-iam");
const util_1 = require("./util");
/**
 * Notify an existing Event Bus of an event.
 *
 * @stability stable
 */
class EventBus {
    /**
     * @stability stable
     */
    constructor(eventBus, props = {}) {
        this.eventBus = eventBus;
        this.role = props.role;
    }
    /**
     * Returns the rule target specification.
     *
     * NOTE: Do not use the various `inputXxx` options. They can be set in a call to `addTarget`.
     *
     * @stability stable
     */
    bind(rule, id) {
        var _b;
        if (this.role) {
            this.role.addToPrincipalPolicy(this.putEventStatement());
        }
        const role = (_b = this.role) !== null && _b !== void 0 ? _b : util_1.singletonEventRole(rule, [this.putEventStatement()]);
        return {
            id: id !== null && id !== void 0 ? id : '',
            arn: this.eventBus.eventBusArn,
            role,
        };
    }
    putEventStatement() {
        return new iam.PolicyStatement({
            actions: ['events:PutEvents'],
            resources: [this.eventBus.eventBusArn],
        });
    }
}
exports.EventBus = EventBus;
_a = JSII_RTTI_SYMBOL_1;
EventBus[_a] = { fqn: "@aws-cdk/aws-events-targets.EventBus", version: "1.94.1" };
